### PR TITLE
lite-apiserver async cache response

### PR DESCRIPTION
**Pull request by Beckham007**
lite-apiserver: async cache response

### CODE REVIEW

**Review by neweyes — COMMENTED** [review]
/lgtm

File: pkg/lite-apiserver/proxy/proxy.go [hunk 1]
```diff
@@ -160,13 +162,23 @@ func (p *EdgeReverseProxy) newTransport() *EdgeTransport {
...
+func (p *EdgeReverseProxy) modifyResponse(resp *http.Response) error {
 	if !p.needCache {
 		return nil
 	}
 
 	// cache response data
-	p.writeCache(NewEdgeResponseDataHolder(res))
+	dupReader, pipeReader := NewDupReadCloser(resp.Body)
+
+	go func(header http.Header, statusCode int, pipeReader io.ReadCloser) {
```
[the remark this answers is not in the document]
Comment (dixudx) [reply pkg/lite-apiserver/proxy/proxy.go:173]: Agree +1.

[the remark this answers is not in the document]
Comment (dixudx) [reply pkg/lite-apiserver/proxy/proxy.go:173]: Actually I'd prefer below way,

```go
pr, pw := io.Pipe()
io.Copy(
	pw,
	io.TeeReader(
		resp.Body,
		chanWriter,
	),
)
resp.Body = pr
```

The `chanWriter` will write body bytes to a buffer-sized string channel. And Another global goroutine will store & update to local cache.

File: pkg/lite-apiserver/proxy/proxy.go [hunk 2]
```diff
@@ -17,7 +17,9 @@ limitations under the License.
 package proxy
 
 import (
+	"bytes"
 	"fmt"
+	"io"
 	apirequest "k8s.io/apiserver/pkg/endpoints/request"
```
Comment (dixudx) [inline pkg/lite-apiserver/proxy/proxy.go:23]: L23 is outstanding. Maybe we could format it by adding a newline.

Comment (Beckham007) [reply pkg/lite-apiserver/proxy/proxy.go]: done. thx

File: pkg/lite-apiserver/proxy/proxy.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Beckham007) [reply pkg/lite-apiserver/proxy/proxy.go:173]: Thx, @neweyes @dixudx. Your suggestions are very good.
This pr forced on fixing http response hanging. 
The CacheManager is very simple now, let's create a new issue to discussing how to cache more effectively? e.g. producer & consumer mode, pipe io, cache context...

[the remark this answers is not in the document]
Comment (dixudx) [reply pkg/lite-apiserver/proxy/proxy.go:173]: Ok, let's move this discussion to a new issue.

**Review by dixudx — APPROVED** [review]
/lgtm

**Pull request merged**